### ISSUE TITLE
Enhancement: A template change should force a new resource

### ISSUE THREAD

**Issue by jason-azze** [terraform-providers/terraform-provider-vsphere#182]
tl;dr: I'd like for vSphere templates to become first-class citizens like AWS AMIs. Right now, the vsphere provider only loves them for their disks.

### Terraform Version
Terraform v0.10.6

### Affected Resource(s)
- vsphere_virtual_machine

### Story
Imagine a workflow where you use Packer to create versioned vSphere templates (in the same way that you might create custom, versioned AMIs). 

Let's say MyTemplate.v1, which has only a single VMWare virtual disk, becomes MyTemplate.v2 when it gets some OS patches and a new version of our web application. Now I want to destroy 1000 VMs that were based on v1 and replace them with new VMs based on v2.

Now imagine a module with an input for template choice. In variables.tf we have:
```hcl
variable "template_choice" {
  description = "Path to the template you want to use. e.g. Templates/MyTemplate.vNN"
}
```
And when we use the module, it's something like:
```hcl
module "linux-build" {
  source                   = "git::http://path.to.myscm/tf-modules.git//vsphere-mod?ref=v0.0.4"
  vsphere_user             = "${var.vsphere_user}"
  vsphere_password         = "${var.vsphere_password}"
  vsphere_server           = "${var.vsphere_server}"
  template_choice          = "Templates/MyTemplate.v2"
  vm_folder_name           = "${var.vm_folder_name}"
  vcpu_count               = "1"
  memory_amount            = "1024"
  dns_domain               = "${var.dns_domain}"
  dns_search_zones         = "${var.dns_search_zones}"
  dns_server_addresses     = "${var.dns_server_addresses}"
  vm_datacenter_name       = "${var.vm_datacenter_name}"
  vm_cluster_name          = "${var.vm_cluster_name}"
  resource_pool_designator = "${var.resource_pool_designator}"
  network_label_designator = "${var.network_label_designator}"
  datastore_designator     = "${var.datastore_designator}"
}
```
Under the immutable infrastructure model, I'd like to have a template version change (or any template name change, of course) cause a destruction/creation event (AKA force a new resource). But because the concept of "template" in the vSphere Provider is currently only concerned with the _disk_ attached to the template, instead TF tries to "update in place". 

This means it tries to yank the boot disk out of a running VM and replace it with a new one from the updated template. This, as you would expect, fails. (I work around this by using taint, but it feels like a kluge.)

Here is the plan:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ module.linux-test.vsphere_virtual_machine.generic
      disk.1613364280.bootable:        "false" => "false"
      disk.1613364280.controller_type: "scsi" => ""
      disk.1613364280.datastore:       "StorageCluster/datastore1" => ""
      disk.1613364280.iops:            "0" => "0"
      disk.1613364280.keep_on_remove:  "false" => "false"
      disk.1613364280.name:            "" => ""
      disk.1613364280.size:            "0" => "0"
      disk.1613364280.template:        "Templates/MyTemplate.v1" => ""
      disk.1613364280.type:            "thin" => ""
      disk.1613364280.vmdk:            "" => ""
      disk.2326733146.bootable:        "" => ""
      disk.2326733146.controller_type: "" => "scsi"
      disk.2326733146.datastore:       "" => "StorageCluster/datastore1"
      disk.2326733146.iops:            "" => ""
      disk.2326733146.keep_on_remove:  "" => ""
      disk.2326733146.key:             "" => <computed>
      disk.2326733146.name:            "" => ""
      disk.2326733146.size:            "" => ""
      disk.2326733146.template:        "" => "Templates/MyTemplate.v2"
      disk.2326733146.type:            "" => "thin"
      disk.2326733146.uuid:            "" => <computed>
      disk.2326733146.vmdk:            "" => ""


Plan: 0 to add, 1 to change, 0 to destroy.

------------------------------------------------------------------------

```
And here is the apply:
```
Error applying plan:

1 error(s) occurred:

* module.linux-test.vsphere_virtual_machine.generic: 1 error(s) occurred:

* vsphere_virtual_machine.generic: [ERROR] Update Remove Disk - Error removing disk: ServerFaultCode: 
Required property device is missing from data object of type VirtualDeviceConfigSpec

while parsing serialized DataObject of type vim.vm.device.VirtualDeviceSpec
at line 2, column 150

while parsing property "deviceChange" of static type ArrayOfVirtualDeviceConfigSpec

while parsing serialized DataObject of type vim.vm.ConfigSpec
at line 2, column 144

while parsing call information for method ReconfigVM_Task
at line 2, column 66

while parsing SOAP body
at line 2, column 60

while parsing SOAP envelope
at line 2, column 0

while parsing HTTP request for method reconfigure
on object of type vim.VirtualMachine
at line 1, column 0
```
Thanks for reading.

**Comment by vancluever** [comment]
Hey 	@jason-azze - just wanted to drop in and say that this issue more sounds like a bug to me - not saying that it's supported by design currently, but it technically should be and considering that I want to make sure it's classified correctly.

We will be reviewing this in the next few weeks while we focus on the `vsphere_virtual_machine` resource.

Thanks!

**Comment by pryorda** [comment]
In my opinion the destruction because of template change should be a flag. You may not always need to destroy a machine to get the same changes that are occuring to the template. Think chef or puppet enforcing a policy.

**Comment by rismoney** [comment]
How about to determine a template has changed, leverage an annotation to store the template name used.  Does template actually mean anything on an instantiated guest anyway?  The concept of originating ami/template isn't there, but could be...

**Comment by jessebolson** [comment]
This is exactly my use-case, and I'd love to see template changes cause a destroy/create as well - if any testing or documentation needs to be done I'd be happy to assist.

**Comment by nowhammies** [comment]
agree I have this use case as well. +1

**Comment by javefang** [comment]
Agreed and this is our use case. We do immutable infrastructure, meaning we always build a new template using Packer when change is required (e.g. upgrading binary version) Currently changing the template does show change but Terraform does not seem to behave correctly.

**Comment by jason-azze** [comment]
Here's a post 1.1.0 update.
I successfully provisioned some VMs by cloning MyTemplate.v1.
I then changed the value of `name` in my templatesource to MyTemplate.v2 by changing the value of my variable, `template_choice`.
```
data "vsphere_virtual_machine" "source-template" {
  name          = "${var.template_choice}"
  datacenter_id = "${data.vsphere_datacenter.target-datacenter.id}"
}
```
I then ran a new plan, and Terraform didn't want to destroy or change anything. This is an improvement in one sense. It didn't try to change the OS disk of a running VM. But a new resource isn't created either.

Because I was just testing, I only changed the name of the template using the vSphere web client "rename" function. Perhaps that isn't enough to change the unique identifier of the template, so TF doesn't think anything has changed. I'll test that.

**Comment by vancluever** [comment]
@jason-azze, you are 100% correct here:

> Because I was just testing, I only changed the name of the template using the vSphere web client "rename" function. Perhaps that isn't enough to change the unique identifier of the template, so TF doesn't think anything has changed. I'll test that.

By doing this you are still using the same template technically, so the UUID will not change. You need to select a different template altogether.

**Comment by bioxcession** [comment]
Awesome! I didn't realize this has been implemented. I will give it a try tomorrow.

**Comment by bill-rich** [comment]
Hi All! The reworking of the virtual machine resource in December appears to have addressed this issue. Now, when the template UUID changes, the virtual machine will be recreated.
